### PR TITLE
[Impersonation] Issue Access Token Exchange for Subject token

### PR DESCRIPTION
**Pull request by Thumimku**
Public Issue: https://github.com/wso2/product-is/issues/20066

## Purpose
When the impersonator acquired subject token, he/she can exchange the token for AT to access resources. This fix will deliver above exchange capability.

## Approach
From the request we validate whether the request is for impersonated access token or not. If yes, then we validate subject token and actor toke. After successful validation, set subject as authorised user and create an access token accoding to delegation spec.  Following attribute will be set in token context.

```
IMPERSONATING_REQUEST -> true
IMPERSONATING_ACTOR -> impersonator
IMPERSONATED_SUBJECT -> subject
```
OAuth Token Issuer will handle the rest.

## Sample Request

```
curl -X POST \
  https://localhost:9443/oauth2/token \
  -H 'Authorization: Basic  <xxx:xxx>' \
  -d 'subject_token=<SUBJECT_TOKEN>' \
  -d 'subject_token_type=urn:ietf:params:oauth:token-type:jwt' \
  -d 'actor_token=<ID_TOKEN>' \
  -d 'actor_token_type=urn:ietf:params:oauth:token-type:id_token' \
  -d 'grant_type=urn:ietf:params:oauth:grant-type:token-exchange' 
```

## Sample Impersonated JWT AT Payload

```
{
  "sub": "8122e3de-0f3b-4b0e-a43a-d0c237451b7a",
  "aut": "APPLICATION_USER",
  "iss": "https://localhost:9443/oauth2/token",
  "client_id": "xnygcXs9Z4L5fhhfDY9MCcnUwxQa",
  "aud": "xnygcXs9Z4L5fhhfDY9MCcnUwxQa",
  "nbf": 1713770236,
  "act": {
    "sub": "d9982d93-4e73-4565-b7ac-3605e8d05f80"
  },
  "azp": "xnygcXs9Z4L5fhhfDY9MCcnUwxQa",
  "org_id": "10084a8d-113f-4211-a0d5-efe36b082211",
  "scope": "internal_login internal_user_mgt_list internal_user_mgt_view openid",
  "exp": 1713773836,
  "org_name": "Super",
  "iat": 1713770236,
  "jti": "9074d48b-0997-4732-8d00-5262f472c139"
}
```

## Demo

https://github.com/wso2-extensions/identity-oauth2-grant-token-exchange/assets/25488962/6f4e9266-63bd-4e07-b82d-df0e4eacfe2a

Merge Condition
- [x] OAuth PR merged
- [x] OAuth version bumped
- [x] Unit test covered. Negative case also covered.

### CODE REVIEW

File: component/org.wso2.carbon.identity.oauth2.grant.token.exchange/src/main/java/org/wso2/carbon/identity/oauth2/grant/token/exchange/TokenExchangeGrantHandler.java [hunk 1]
```diff
@@ -136,6 +155,219 @@ public boolean validateGrant(OAuthTokenReqMessageContext tokReqMsgCtx) throws Id
...
+     * @param tokReqMsgCtx
+     * @return An array of individual scope strings extracted from the scopes claim.
+     */
+    private String[] getScopes(JWTClaimsSet claimsSet, OAuthTokenReqMessageContext tokReqMsgCtx) {
+
+        String[] requestedScopes = tokReqMsgCtx.getScope();
+        String[] approvedScopes = ((String) claimsSet.getClaims().get(SCOPE)).split("\\s+");
+        if (requestedScopes == null) {
+            return approvedScopes;
+        } else {
```
Comment (DMHP) [inline component/org.wso2.carbon.identity.oauth2.grant.token.exchange/src/main/java/org/wso2/carbon/identity/oauth2/grant/token/exchange/TokenExchangeGrantHandler.java:247]: else is not needed.
 String[] requestedScopes = tokReqMsgCtx.getScope();
        String[] approvedScopes = ((String) claimsSet.getClaims().get(SCOPE)).split("\\s+");
        if (requestedScopes == null) {
            return approvedScopes;
        } 
return filterRequestedScopes(requestedScopes, approvedScopes);
}

File: component/org.wso2.carbon.identity.oauth2.grant.token.exchange/src/main/java/org/wso2/carbon/identity/oauth2/grant/token/exchange/TokenExchangeGrantHandler.java [hunk 2]
```diff
@@ -136,6 +155,219 @@ public boolean validateGrant(OAuthTokenReqMessageContext tokReqMsgCtx) throws Id
...
+            return filterRequestedScopes(requestedScopes, approvedScopes);
+        }
+    }
+
+    private String[] filterRequestedScopes(String[] requestedScopes, String[] approvedScopes) {
+
+        Set<String> approvedScopesSet = new HashSet<>(Arrays.asList(approvedScopes));
+        Set<String> commonScopes = new HashSet<>();
+        for (String element : requestedScopes) {
```
Comment (DMHP) [inline component/org.wso2.carbon.identity.oauth2.grant.token.exchange/src/main/java/org/wso2/carbon/identity/oauth2/grant/token/exchange/TokenExchangeGrantHandler.java:256]: Add a meaningful name for element.

File: component/org.wso2.carbon.identity.oauth2.grant.token.exchange/src/main/java/org/wso2/carbon/identity/oauth2/grant/token/exchange/utils/TokenExchangeUtils.java [hunk 3]
```diff
@@ -342,6 +345,50 @@ public static void setAuthorizedUser(OAuthTokenReqMessageContext tokenReqMsgCtx,
...
+
+            // Populate IDP groups attribute
+            populateIdPGroupsAttribute(tokenReqMsgCtx, identityProvider, claimsSet);
+        } catch (UserStoreException e) {
+            // Handle user store exception
+            throw new IdentityOAuth2Exception(OAuth2ErrorCodes.INVALID_REQUEST,
+                    "Failed to resolve username from authenticated subject identifier", e);
+        }
+    }
+
```
Comment (DMHP) [inline component/org.wso2.carbon.identity.oauth2.grant.token.exchange/src/main/java/org/wso2/carbon/identity/oauth2/grant/token/exchange/utils/TokenExchangeUtils.java:390]: Additional new line

Comment (Thumimku) [reply component/org.wso2.carbon.identity.oauth2.grant.token.exchange/src/main/java/org/wso2/carbon/identity/oauth2/grant/token/exchange/utils/TokenExchangeUtils.java:390]: resolved

File: component/org.wso2.carbon.identity.oauth2.grant.token.exchange/src/main/java/org/wso2/carbon/identity/oauth2/grant/token/exchange/TokenExchangeGrantHandler.java (hunk 2 above)
Comment (Thumimku) [reply component/org.wso2.carbon.identity.oauth2.grant.token.exchange/src/main/java/org/wso2/carbon/identity/oauth2/grant/token/exchange/TokenExchangeGrantHandler.java]: resolved in https://github.com/wso2-extensions/identity-oauth2-grant-token-exchange/pull/26/commits/d1762473ae0c89124fb4cd7a6158649e542a595f

File: component/org.wso2.carbon.identity.oauth2.grant.token.exchange/src/main/java/org/wso2/carbon/identity/oauth2/grant/token/exchange/TokenExchangeGrantHandler.java (hunk 1 above)
Comment (Thumimku) [reply component/org.wso2.carbon.identity.oauth2.grant.token.exchange/src/main/java/org/wso2/carbon/identity/oauth2/grant/token/exchange/TokenExchangeGrantHandler.java]: +1 ref: https://github.com/wso2-extensions/identity-oauth2-grant-token-exchange/pull/26/commits/d1762473ae0c89124fb4cd7a6158649e542a595f

**Review by DMHP — APPROVED** [review]

**Pull request merged**